### ISSUE TITLE
Perf: Postpone file system calls during startup

### ISSUE THREAD

**Issue by alexdima** [microsoft/vscode#169868]
* apply the patch that logs fs calls which is included below
* run `./scripts/code.sh`
* open a folder and open a single css file
* don't modify it (i.e. avoid hot exit) and reload the window

<details><summary>Patch</summary>

```diff
diff --git a/src/vs/editor/browser/view.ts b/src/vs/editor/browser/view.ts
index f912513d25f..91da47ff4f8 100644
--- a/src/vs/editor/browser/view.ts
+++ b/src/vs/editor/browser/view.ts
@@ -379,6 +379,7 @@ export class View extends ViewEventHandler {
 		if (!dom.isInDOM(this.domNode.domNode)) {
 			return;
 		}
+		console.warn(`RENDERING EDITOR!`);
 
 		let viewPartsToRender = this._getViewPartsToRender();
 
diff --git a/src/vs/platform/files/common/diskFileSystemProviderClient.ts b/src/vs/platform/files/common/diskFileSystemProviderClient.ts
index 4eeed662440..9a9807ad296 100644
--- a/src/vs/platform/files/common/diskFileSystemProviderClient.ts
+++ b/src/vs/platform/files/common/diskFileSystemProviderClient.ts
@@ -72,10 +72,12 @@ export class DiskFileSystemProviderClient extends Disposable implements
 	//#region File Metadata Resolving
 
 	stat(resource: URI): Promise<IStat> {
+		console.log(`stat: ${resource.toString()}`);
 		return this.channel.call('stat', [resource]);
 	}
 
 	readdir(resource: URI): Promise<[string, FileType][]> {
+		console.log(`readdir: ${resource.toString()}`);
 		return this.channel.call('readdir', [resource]);
 	}
 
@@ -84,12 +86,14 @@ export class DiskFileSystemProviderClient extends Disposable implements
 	//#region File Reading/Writing
 
 	async readFile(resource: URI, opts?: IFileAtomicReadOptions): Promise<Uint8Array> {
+		console.log(`readFile: ${resource.toString()}`);
 		const { buffer } = await this.channel.call('readFile', [resource, opts]) as VSBuffer;
 
 		return buffer;
 	}
 
 	readFileStream(resource: URI, opts: IFileReadStreamOptions, token: CancellationToken): ReadableStreamEvents<Uint8Array> {
+		console.log(`readFileStream: ${resource.toString()}`);
 		const stream = newWriteableStream<Uint8Array>(data => VSBuffer.concat(data.map(data => VSBuffer.wrap(data))).buffer);
 
 		// Reading as file stream goes through an event to the remote side
@@ -148,6 +152,7 @@ export class DiskFileSystemProviderClient extends Disposable implements
 	}
 
 	writeFile(resource: URI, content: Uint8Array, opts: IFileWriteOptions): Promise<void> {
+		console.log(`writeFile: ${resource.toString()}`);
 		return this.channel.call('writeFile', [resource, VSBuffer.wrap(content), opts]);
 	}
 
@@ -180,6 +185,7 @@ export class DiskFileSystemProviderClient extends Disposable implements
 	//#region Move/Copy/Delete/Create Folder
 
 	mkdir(resource: URI): Promise<void> {
+		console.log(`mkdir: ${resource.toString()}`);
 		return this.channel.call('mkdir', [resource]);
 	}
 
```

</details>

Observe the following fs calls that occur before the first editor rendering:
```
      readFile: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json
      readFile: file:///Users/alex/src/working_dir/.vscode/settings.json
      readFile: file:///Users/alex/src/working_dir/.vscode/tasks.json
      readFile: file:///Users/alex/src/working_dir/.vscode/launch.json
          stat: file:///Users/alex/src/working_dir/.vscode/settings.json
          stat: file:///Users/alex/src/working_dir/.vscode/tasks.json
          stat: file:///Users/alex/src/working_dir/.vscode/launch.json
          stat: file:///Users/alex/src/working_dir
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/Backups/2254581727a54bbc3c426f0598ad1144
      readFile: file:///Users/alex/Library/Application%20Support/code-oss-dev/installSource
      readFile: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/keybindings.json
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/installSource
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/keybindings.json
          stat: file:///Users/alex/src/working_dir
readFileStream: file:///Users/alex/src/working_dir/test.css
          stat: file:///Users/alex/src/working_dir/test.css
readdir: file:///Users/alex/src/working_dir
          stat: file:///Users/alex/src/vscode/extensions
      readFile: file:///Users/alex/.vscode-oss-dev/extensions/control.json
          stat: file:///Users/alex/.vscode-oss-dev/extensions/extensions.json
          stat: file:///Users/alex/.vscode-oss-dev/extensions/control.json
      readFile: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json
      readFile: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/tasks.json
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/User/tasks.json
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/window1/output_20221222T180040
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/userDataSync.log
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/editSessions.log
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/window1/renderer.log
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/main.log
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/sharedprocess.log
          stat: file:///Users/alex/Library/Application%20Support/code-oss-dev/logs/20221222T153340/ptyhost.log
          stat: file:///Users/alex/src/vscode/extensions
      readFile: file:///Users/alex/.vscode-oss-dev/extensions/extensions.json
          stat: file:///Users/alex/.vscode-oss-dev/extensions/extensions.json
          stat: file:///Users/alex/src/vscode/.build/builtInExtensions/ms-vscode.js-debug-companion
          stat: file:///Users/alex/src/vscode/.build/builtInExtensions/ms-vscode.js-debug
          stat: file:///Users/alex/src/vscode/.build/builtInExtensions/ms-vscode.vscode-js-profile-table
RENDERING EDITOR!
```

IMHO the following fs calls should be postponed:
- [ ] `file:///Users/alex/src/working_dir/.vscode/tasks.json` @sandy081 
- [ ] `file:///Users/alex/src/working_dir/.vscode/launch.json` @sandy081 
- [ ] `file:///Users/alex/Library/Application%20Support/code-oss-dev/User/tasks.json` @sandy081 
- [x] `file:///Users/alex/Library/Application%20Support/code-oss-dev/installSource` @joaomoreno  

I also noticed that we read the following files twice:
- [ ] `file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json` @sandy081

**Comment by bpasero** [comment]
I have no clue what we use install source for, but lots of digging reveals this commit: https://github.com/microsoft/vscode/commit/14699ba54eb57e73eb61a503421659eb04f41749#diff-ce90ef6ffd304cc541efa8135dc4878c7b61889d2b2d5ef69e196af6cea7647b

**Comment by sandy081** [comment]
> `file:///Users/alex/src/working_dir/.vscode/tasks.json @sandy081`
`file:///Users/alex/src/working_dir/.vscode/launch.json`

These are part of the configuration model and is sent to the extension host on start up. Best is that these can be postponed to the point where we need them, that is, starting extension host. Not sure if that helps with the start up perf?

**Comment by sandy081** [comment]
Updating the list - `file:///Users/alex/Library/Application%20Support/code-oss-dev/User/tasks.json` is only read once.

**Comment by sandy081** [comment]
`file:///Users/alex/Library/Application%20Support/code-oss-dev/User/settings.json` is read twice because

- First time during initialising configuration without user tasks
- User tasks is read later when there is an idle time - by reloading user configuration - during which settings and tasks files are read.

**Comment by joaomoreno** [comment]
Install source is a common telemetry property. It can be set after installing VS Code. It came as an ask from Anaconda, to know how many installations came from them. We decided to go with creating a new file rather than updating existing files, to avoid scenarios where we'd break important files (eg package.json or product.json) Here's the latest telemetry numbers:

![image](https://user-images.githubusercontent.com/22350/210373691-3fdcdb2b-113e-4397-8d69-8c8966b055fa.png)

**Comment by joaomoreno** [comment]
`installSource` is gone! https://github.com/microsoft/vscode/pull/176688

**Comment by sandy081** [comment]
Tasks and Launches are part of Config and they are accessed sync from config service. In order to prevent reading these files during start up we should 

1. Delay loading them until they are needed. (probably workbench restored phase)
2. Have an assertion to make sure no component is reading them until then

**Comment by sandy081** [comment]
> Tasks and Launches are part of Config and they are accessed sync from config service

This needs considerable refactoring and I do not think it is worth a change for the gain given that these files are resolved in parallel.  Hence marking this as no fix.
